### ISSUE TITLE
Exception running project with 3.4.2

### ISSUE THREAD

**Issue by frangarcia** [vert-x3/vertx-lang-groovy#51]
Opened here from https://github.com/eclipse/vert.x/issues/2022

I was just running a vertx project we have with the new 3.4.2 version and I got this exception:
```bash
startup failed:
General error during class generation: io.vertx.groovy.core.VertxPropertiesModuleFactory cannot be cast to org.codehaus.groovy.runtime.m12n.PropertiesModuleFactory

java.lang.ClassCastException: io.vertx.groovy.core.VertxPropertiesModuleFactory cannot be cast to org.codehaus.groovy.runtime.m12n.PropertiesModuleFactory
at org.codehaus.groovy.runtime.m12n.StandardPropertiesModuleFactory.newModule(StandardPropertiesModuleFactory.java:41)
at org.codehaus.groovy.runtime.m12n.ExtensionModuleScanner.scanExtensionModuleFromProperties(ExtensionModuleScanner.java:79)
at org.codehaus.groovy.runtime.m12n.ExtensionModuleScanner.scanExtensionModuleFromMetaInf(ExtensionModuleScanner.java:74)
at org.codehaus.groovy.runtime.m12n.ExtensionModuleScanner.scanClasspathModules(ExtensionModuleScanner.java:56)
at org.codehaus.groovy.transform.stc.StaticTypeCheckingSupport$ExtensionMethodCache.getExtensionMethods(StaticTypeCheckingSupport.java:1984)
at org.codehaus.groovy.transform.stc.StaticTypeCheckingSupport.findDGMMethodsForClassNode(StaticTypeCheckingSupport.java:208)
at org.codehaus.groovy.transform.stc.StaticTypeCheckingSupport.findDGMMethodsForClassNode(StaticTypeCheckingSupport.java:194)
at org.codehaus.groovy.transform.stc.StaticTypeCheckingSupport.findDGMMethodsByNameAndArguments(StaticTypeCheckingSupport.java:923)
at org.codehaus.groovy.transform.stc.StaticTypeCheckingSupport.findDGMMethodsByNameAndArguments(StaticTypeCheckingSupport.java:910)
at org.codehaus.groovy.transform.sc.transformers.BooleanExpressionTransformer$OptimizingBooleanExpression.visit(BooleanExpressionTransformer.java:155)
at org.codehaus.groovy.classgen.asm.StatementWriter.writeIfElse(StatementWriter.java:284)
at org.codehaus.groovy.classgen.AsmClassGenerator.visitIfElse(AsmClassGenerator.java:586)
at org.codehaus.groovy.ast.stmt.IfStatement.visit(IfStatement.java:43)
at org.codehaus.groovy.classgen.asm.StatementWriter.writeBlockStatement(StatementWriter.java:85)
at org.codehaus.groovy.classgen.asm.sc.StaticTypesStatementWriter.writeBlockStatement(StaticTypesStatementWriter.java:65)
at org.codehaus.groovy.classgen.AsmClassGenerator.visitBlockStatement(AsmClassGenerator.java:570)
at org.codehaus.groovy.ast.stmt.BlockStatement.visit(BlockStatement.java:71)
at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitClassCodeContainer(ClassCodeVisitorSupport.java:104)
at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitConstructorOrMethod(ClassCodeVisitorSupport.java:115)
at org.codehaus.groovy.classgen.AsmClassGenerator.visitStdMethod(AsmClassGenerator.java:434)
at org.codehaus.groovy.classgen.AsmClassGenerator.visitConstructorOrMethod(AsmClassGenerator.java:387)
at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitConstructor(ClassCodeVisitorSupport.java:122)
at org.codehaus.groovy.classgen.AsmClassGenerator.visitConstructor(AsmClassGenerator.java:506)
at org.codehaus.groovy.ast.ClassNode.visitContents(ClassNode.java:1077)
at org.codehaus.groovy.ast.ClassCodeVisitorSupport.visitClass(ClassCodeVisitorSupport.java:53)
at org.codehaus.groovy.classgen.AsmClassGenerator.visitClass(AsmClassGenerator.java:233)
at org.codehaus.groovy.control.CompilationUnit$17.call(CompilationUnit.java:825)
at org.codehaus.groovy.control.CompilationUnit.applyToPrimaryClassNodes(CompilationUnit.java:1065)
at org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:603)
at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:581)
at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:558)
at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:537)
at org.gradle.api.internal.tasks.compile.ApiGroovyCompiler.execute(ApiGroovyCompiler.java:167)
at org.gradle.api.internal.tasks.compile.ApiGroovyCompiler.execute(ApiGroovyCompiler.java:54)
at org.gradle.api.internal.tasks.compile.daemon.CompilerDaemonServer.execute(CompilerDaemonServer.java:55)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
at org.gradle.messaging.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:364)
at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
at java.lang.Thread.run(Thread.java:748)

1 error
```

Is there any breaking change in the new version?

**Comment by frangarcia** [comment]
Found this issue in the groovy project: https://issues.apache.org/jira/browse/GROOVY-8173

**Comment by vietj** [comment]
good to know thanks

**Comment by abcfy2** [comment]
I have the same issue when build with gradle，here is my build.gradle:
```
buildscript {
    repositories {
        mavenLocal()
        jcenter()
        mavenCentral()
    }
    dependencies {
        classpath 'com.github.jengelman.gradle.plugins:shadow:2.0.0'
    }
}

apply plugin: 'groovy'
apply plugin: 'com.github.johnrengelman.shadow'

if (!JavaVersion.current().java8Compatible) {
    throw new IllegalStateException('''Please install Java 8!'''.stripMargin())
}

sourceSets.main.java.srcDirs = []
sourceSets.main.groovy.srcDirs += ["src/main/java"]

repositories {
    mavenLocal()
    jcenter()
    mavenCentral()
}

dependencies {
    compile "org.codehaus.groovy:groovy-all:${GROOVY_VER}"
    compile "io.vertx:vertx-core:${VERTX_VER}"
    compile "io.vertx:vertx-web:${VERTX_VER}"
    compile "io.vertx:vertx-auth-jwt:${VERTX_VER}"
    compile "io.vertx:vertx-circuit-breaker:${VERTX_VER}"
    compile "io.vertx:vertx-ignite:${VERTX_VER}"
    compile "ch.qos.logback:logback-classic:${LOGBACK_VER}"

    testCompile "org.spockframework:spock-core:${SPOCK_VER}"
}

version = '0.1'

shadowJar {
    classifier = 'fat'
    baseName = 'dgate'
    manifest {
        attributes 'Main-Class': 'top.dteam.dgate.Launcher'
        attributes 'Main-Verticle': 'top.dteam.dgate.MainVerticle'
    }
    mergeServiceFiles {
        include 'META-INF/services/io.vertx.core.spi.VerticleFactory'
    }
}
```

And here is my gradle.properties:
```
GROOVY_VER=2.4.12
SPOCK_VER=1.1-groovy-2.4
VERTX_VER=3.4.2
LOGBACK_VER=1.2.3
```

**Comment by abcfy2** [comment]
Reproduce steps:
1. clone source code: `git clone --depth 1 https://github.com/dteam-top/dgate`
2. `cd dgate && gradle test`(Must install gradle before), test will be pass.
3. Upgrade to vertx 3.4.2:
    1. Modify `gradle.properties` set `VERTX_VER = 3.4.2`
    2. Change `io.vertx.ext.auth.jwt.JWT` to `io.vertx.ext.jwt.JWT` according break changes. `sed -i 's/io.vertx.ext.auth.jwt.JWT/io.vertx.ext.jwt.JWT/g' src/main/java/top/dteam/dgate/utils/JWTTokenRefresher.java b/src/main/java/top/dteam/dgate/utils/JWTTokenRefresher.java src/main/java/top/dteam/dgate/utils/Utils.java b/src/main/java/top/dteam/dgate/utils/Utils.java src/main/java/top/dteam/dgate/handler/JWTTokenSniffer.java b/src/main/java/top/dteam/dgate/handler/JWTTokenSniffer.java`
    3. `gradle run` again

**Comment by jojopad** [comment]
I encountered the compilation issue also with a Groovy project that uses Vertx. After removing the @CompileStatic usage on some Groovy classes, the compilation issue was gone.

**Comment by vietj** [comment]
fixed in 3.5.0

**Comment by vietj** [comment]
please reopen if you see it with 3.5.0
